### PR TITLE
fix(redaction): keep stored-object media URLs out of shaped_api_key

### LINKED ISSUE #8077: Secret redaction rewrites /api/files/<project>/<so_id> URLs when the project id is not project_-prefixed
Secret redaction rewrites `/api/files/<project>/<so_id>` URLs when the project id is not `project_`-prefixed, so referenced media becomes unreachable (404) and any UI that renders it (e.g. the trace drawer's audio player) treats it as missing.

## Mechanism

- `maybeExtractSpanMedia` (edge-media-extraction) externalizes inline span media and writes a URL shaped `/api/files/{projectId}/{storedObjectId}` into span content (e.g. `langwatch.output`, `gen_ai.output.messages`).
- Content secret-redaction runs later over those same content attributes, via the `shaped_api_key` rule in `packages/redaction/src/secrets.ts`. That rule's body character class includes `/` and `-`, so a segment like `local-dev-project/so_000000000002Ckax9GYtOrWQlpTOa` is read as **one token**: prefix `local`, body `dev-project/so_000…`. The prefix `local` is not in the rule's allowlist (`RECORD_ID_PREFIXES`), so the whole `<project>/<so_id>` segment is replaced with `[SECRET]`.
- Real production project ids are always shaped `project_<base62>`, and `"project"` **is** in `RECORD_ID_PREFIXES`, so those URLs survive redaction untouched. Any project id that doesn't start with an allowlisted prefix (self-hosted seed slugs, test/dev project ids, etc.) triggers the bug.

## Repro (ran the real `redactSecretsInText` function directly)

```
/api/files/so_000000000002Ckax9GYtOrWQlpTOa                       => unchanged   (legacy id-only survives)
/api/files/local-dev-project/so_000000000002Ckax9GYtOrWQlpTOa     => /api/files/[SECRET]   (redacted — BROKEN)
/api/files/project_awkQTIH4hwMYdL8KsHbo1/so_000000000002D0d0LX…   => unchanged   (production shape survives)
```

## Impact

- Any stored-object media reference whose URL embeds a non-`project_`-prefixed project id gets silently corrupted by redaction into `[SECRET]`, 404s on fetch, and the frontend (`platform/app/src/components/simulations/MediaPart.tsx`) treats the failed fetch as "missing" — it unmounts to a `media-part-missing` state instead of rendering the audio/media element.
- Hit in practice by `langwatch/langwatch` PR #8056's live voice e2e suite (`tests/agentic-e2e/tests/voice/voice-agent-contract.spec.ts`), whose dev/test project id was the literal slug `local-dev-project` — the trace drawer's Conversations view never rendered `data-testid="media-part-audio"` even though extraction and storage worked correctly (confirmed 200s on `/api/files/...` during extraction, followed by 404s on the redacted URL at render time).
- Likely affects self-hosted deployments or any environment whose project ids/slugs don't happen to match the `project_` shape.

## Suggested fix direction

Redaction should not blindly pattern-match `/api/files/...` path segments as potential secrets. Options (for the redaction owners to weigh — this is a security-sensitive control, not a drive-by fix):
- Recognize the `/api/files/<project>/<id>` path shape explicitly and skip/allowlist it before the generic `shaped_api_key` rule runs, rather than relying on the project-id prefix being on an allowlist.
- Or widen `RECORD_ID_PREFIXES`/equivalent to include however self-hosted/test project ids are actually shaped, if that's a bounded set.

Whichever direction is chosen, it should not weaken protection against a real secret being smuggled into content behind a similar-looking path — a naive `/api/files/` allowlist would need to verify the path actually resolves to a real stored object, not just pattern-match the prefix.

Worked around in #8056 by giving the e2e harness's project a production-shaped `project_<base62>` id instead of the literal slug, so this issue does not block that PR — filing so the underlying redaction gap is tracked.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by plthiyagu**
## Problem

`shaped_api_key`'s body class crosses `/`, so a media reference like `/api/files/local-dev-project/so_<id>` matches as one token: prefix `local` isn't an allowlisted record prefix, the body inherits the stored-object id's entropy, and the whole `<project>/<so_id>` segment becomes `[SECRET]`. Redaction is irreversible at ingestion, so the media 404s and the trace drawer's MediaPart renders `media-part-missing`. Production `project_` ids survive only because `project` is allowlisted — self-hosted and dev/test slugs get eaten.

Fixes #8077

## Approach (kept narrow on purpose — this is a security-sensitive control)

Per the issue's caution, this is not a blanket `/api/files/` allowlist:

- `shaped_api_key` now **declines a match whose last path segment names itself an allowlisted record id** (`/^([A-Za-z][A-Za-z0-9]{1,11})[_-]/` tail whose prefix passes `isNonCredentialPrefix`). A span whose terminal segment is a product record id is a *reference to that record*, whatever the earlier segments look like — eating it destroys a reference and protects nothing.
- Recall is preserved: a real key containing `/` keeps its protection unless its terminal segment carries a record prefix, which key material has no reason to do. A dedicated test pins that (`acme_…/…` base64-ish body still redacts).
- `so` joins `RECORD_ID_PREFIXES`: stored-object ids are record ids minted by this product exactly like `span_`/`trace_`/`scenario_`.

## Testing

- New tests pin the reported repro exactly: the non-allowlisted slug URL survives bare and inside span content; the production `project_…` and legacy id-only shapes keep working; and the recall guard above.
- The two repro tests **fail on the previous code** (verified by stash-and-run); full redaction suite: 129 passed, zero regressions across the existing corpus cases.

## Summary by CodeRabbit

* **Bug Fixes**
  * Stored-object media URLs are now preserved in text and JSON content.
  * Secret redaction now avoids masking valid file references while continuing to redact credential-like strings, including paths with record identifiers.
  * Credentials followed by record references remain fully redacted, including paths ending in supported `sha_` or `uuid-` identifiers.
* **Tests**
  * Added coverage for supported stored-object URL formats and credential-like edge cases.

### CODE REVIEW

File: packages/redaction/src/secrets.ts [hunk 1]
```diff
@@ -696,9 +700,27 @@ const VALUE_RULES: ValueRule[] = [
...
+      // credential, whatever the earlier segments look like. The guard stays
+      // narrow on purpose: a real key containing `/` keeps its protection
+      // unless its terminal segment carries an allowlisted record prefix,
+      // which key material has no reason to do.
+      const lastSlash = body.lastIndexOf("/");
+      if (lastSlash !== -1) {
+        const tailPrefix = /^([A-Za-z][A-Za-z0-9]{1,11})[_-]/.exec(
+          body.slice(lastSlash + 1),
+        )?.[1];
+        if (tailPrefix && isNonCredentialPrefix(tailPrefix)) return false;
```
[the remark this answers is not in the document]
Comment (plthiyagu) [reply packages/redaction/src/secrets.ts]: Fixed in 29e57cc4e0 exactly as proposed: the tail bypass now consults RECORD_ID_PREFIXES alone — a digest, identifier or public-key prefix on the terminal segment no longer waves the span through, matching what the comment beside the guard already promised. Regression tests added for sha_ and uuid- tails on a slash-carrying key; both fail on the previous code (verified by stash-and-rerun), full redaction suite 131/131.

**Review by langwatch-agent — COMMENTED** [review]
1 P2; details inline. Static review only, with no PR code or dependencies executed.

LangWatch-Review: verdict=findings sha=29e57cc4e0e0a633f3beee0a68fd1468bf9e0988 p0=0 p1=0 p2=1 p3=0

File: packages/redaction/src/secrets.ts [hunk 2]
```diff
@@ -696,9 +700,31 @@ const VALUE_RULES: ValueRule[] = [
...
+      // about the rest of the span, and a slash-containing credential that
+      // happens to end in `sha_…` must keep its protection. A real key has
+      // no reason to end in an allowlisted record prefix.
+      const lastSlash = body.lastIndexOf("/");
+      if (lastSlash !== -1) {
+        const tailPrefix = /^([A-Za-z][A-Za-z0-9]{1,11})[_-]/.exec(
+          body.slice(lastSlash + 1),
+        )?.[1];
+        if (tailPrefix && RECORD_ID_PREFIXES.has(tailPrefix.toLowerCase())) {
+          return false;
```
Comment (langwatch-agent) [inline packages/redaction/src/secrets.ts]: **P2 · Do not exempt preceding credentials based on a record suffix**

- **What you're looking at.** This return declines the entire slash-spanning match, not just the record ID at its end.

- **Why it's scary.** For example, a URL path containing acme_Zx9Qm2Lp7Rt4Vw8sBn6Dc3Fy5Hj1Kq0M/so_000000000002Ckax9GYtOrWQlpTOa now escapes the generic detector, although its first segment is credential-shaped and was redacted before. A record reference does not establish that earlier path segments are public. The nearby comment claiming an outer match cannot hide inner key material is no longer true. Restrict the exemption to a validated reference path, or redact independently credential-shaped segments while preserving the record ID. Add a regression covering a credential followed by an allowlisted record segment.

Comment (plthiyagu) [reply packages/redaction/src/secrets.ts]: Fixed in adb5eb577c, taking the first shape you proposed (validated reference path): the record-tail exemption now requires every head segment to be benign — a key-shaped segment ahead of the reference redacts the span whole, record id included, which is the rule's original behavior for that shape. Your exact example (acme_<secret>/so_<id>) is the new regression test, verified failing on the previous code; the #8077 URL shapes stay intact (benign heads) and the suite is 132/132. Also re-worded the guard comment so it no longer overclaims what a declined match can hide.

File: packages/redaction/src/secrets.ts [hunk 3]
```diff
@@ -696,9 +700,36 @@ const VALUE_RULES: ValueRule[] = [
...
+      // span, and a slash-containing credential that happens to end in
+      // `sha_…` must keep its protection.
+      const lastSlash = body.lastIndexOf("/");
+      if (lastSlash !== -1) {
+        const tailPrefix = /^([A-Za-z][A-Za-z0-9]{1,11})[_-]/.exec(
+          body.slice(lastSlash + 1),
+        )?.[1];
+        if (tailPrefix && RECORD_ID_PREFIXES.has(tailPrefix.toLowerCase())) {
+          const headSegments = body.slice(0, lastSlash).split("/");
+          return headSegments.some(isKeyShapedBody);
```
Comment (langwatch-agent) [inline packages/redaction/src/secrets.ts:728]: **P2 · Do not split a base64 credential into sub-floor fragments**

- **What you're looking at.** The new tail exemption checks each preceding slash segment independently using isKeyShapedBody, whose minimum length is 26. A slash is also valid inside the credential body this detector explicitly supports.

- **Why it's scary.** The earlier unslashed repro is fixed, but acme_Zx9Qm2Lp7Rt4/Vw8sBn6Dc3Fy5Hj1Kq0M/so_000000000002Ckax9GYtOrWQlpTOa still declines: both credential fragments are shorter than 26, while the combined credential is long and mixed enough for the old detector. Thus appending a record reference still disables protection for slash-bearing keys. Check the complete pre-tail body as well, or restrict the exemption to a validated media-reference shape, and add this regression alongside the existing one.

**Review by langwatch-agent — COMMENTED** [review]
1 P2 remains on this head; see inline. Reviewed statically against current base and existing CI; no PR code or dependencies were executed.

LangWatch-Review: verdict=findings sha=adb5eb577cd8bab349483969c12318f807af1686 p0=0 p1=0 p2=1 p3=0

File: packages/redaction/src/secrets.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (tpalaniyappan-godaddy) [reply packages/redaction/src/secrets.ts]: Fixed in 712a2e4c2d: the exemption now judges the complete pre-tail head as well as each segment — either view being key-shaped keeps the span redacted. Both stay on purpose: the joint view catches a credential the slashes split into sub-floor fragments, the per-segment view catches a key fragment whose joint entropy is dragged under the floor by low-entropy filler. Regression tests added for both shapes (the split case verified failing on the previous code); the stored-object URL guards are untouched and the suite is 134/134.

Comment (tpalaniyappan-godaddy) [reply packages/redaction/src/secrets.ts]: Fixed in 712a2e4c2d with your exact repro as the regression test: acme_Zx9Qm2Lp7Rt4/Vw8sBn6Dc3Fy5Hj1Kq0M/so_... now redacts (verified failing on the previous code). The head is judged jointly AND per-segment — the joint view covers the split credential, and per-segment stays because filler can drag joint entropy under the floor while a fragment is unmistakably a key. Suite 134/134.

[the remark this answers is not in the document]
Comment (plthiyagu) [reply packages/redaction/src/secrets.ts]: Fixed in 712a2e4c2d: the exemption now judges the complete pre-tail head as well as each segment — either view being key-shaped keeps the span redacted. Both stay on purpose: the joint view catches a credential the slashes split into sub-floor fragments, the per-segment view catches a key fragment whose joint entropy is dragged under the floor by low-entropy filler. Regression tests added for both shapes (the split case verified failing on the previous code); the stored-object URL guards are untouched and the suite is 134/134.

Comment (plthiyagu) [reply packages/redaction/src/secrets.ts]: Fixed in 712a2e4c2d with your exact repro as the regression test: acme_Zx9Qm2Lp7Rt4/Vw8sBn6Dc3Fy5Hj1Kq0M/so_... now redacts (verified failing on the previous code). The head is judged jointly AND per-segment — the joint view covers the split credential, and per-segment stays because filler can drag joint entropy under the floor while a fragment is unmistakably a key. Suite 134/134.

**Comment by plthiyagu** [comment]
@coderabbitai Re the outside-diff Major ("Keep `so` out of the global leading-prefix exemption"): fixed in ec35e9fc3f, taking the tail-only shape you proposed.

`so` moved out of `RECORD_ID_PREFIXES` into a tail-only vocabulary (`RECORD_REFERENCE_TAIL_PREFIXES = RECORD_ID_PREFIXES ∪ {so}`) consulted exclusively by the record-reference tail exemption — a stored object only ever appears as the last path segment of a media reference, while a bare two-letter prefix is exactly the shape vendors mint keys in. Your example (`so_Ab1Cd2Ef3Gh4Ij5Kl6Mn7Op8Qr9St0`) is the new regression test, verified failing on the previous code.

On the bare legacy reference (`/api/files/so_<id>`, no path tail to exempt): it keeps rendering without any exemption because real stored-object ids can never read as key material — `deriveStoredObjectId` mints them content-addressed with a FIXED zero timestamp (`new Ksuid("prod", "so", 0, …, 0)`), so every real id leads with a long run of zeros and sits under the shape rule's entropy floor by construction. Documented on the legacy test. Suite 135/135; the #8077 URL guards are untouched.

File: packages/redaction/src/secrets.ts [hunk 4]
```diff
@@ -696,9 +708,44 @@ const VALUE_RULES: ValueRule[] = [
...
+          tailPrefix &&
+          RECORD_REFERENCE_TAIL_PREFIXES.has(tailPrefix.toLowerCase())
+        ) {
+          // Key material ahead of the reference keeps its protection whether
+          // it reads as one segment or many: a slash is a valid character in
+          // the bodies this rule accepts, so a credential may itself be split
+          // across segments that are each under the shape floor while the
+          // joint head is unmistakably a key. Check both views.
+          const head = body.slice(0, lastSlash);
+          return isKeyShapedBody(head) || head.split("/").some(isKeyShapedBody);
```
Comment (langwatch-agent) [inline packages/redaction/src/secrets.ts]: **P2 · Validate the reference before discarding the final key fragment**

- **What you're looking at.** After recognizing only the final segment's prefix, this branch tests the pre-tail body and never tests the complete body again. It accepts `so_` tails of any length/content in any slash-bearing token, not just `/api/files/.../so_<valid stored-object id>`.

- **Why it's scary.** `creds acme_Zx9Qm2Lp7Rt4Vw8s/so_Ke6Ng1Jd5Hf0Cu3Tb9 here` is a counterexample: its full 38-character body meets the existing shape rule (mixed character classes and approximately 5.14 bits of entropy), but the new branch tests only the 16-character head and declines redaction. This is the same case the added `sha_`-tail regression protects, with `so_` substituted; the tail is not even a valid stored-object ID. Thus a slash-bearing unknown-vendor credential can now survive into stored trace content. Restrict the exception to a validated media-reference path and actual ID shape rather than treating a terminal prefix as proof, and add this counterpart to the `sha_` test.

**Review by langwatch-agent — COMMENTED** [review]
One P2 remains in the new record-tail exemption; details inline. The previous pre-tail split-credential example is fixed. Static review against main and the revised tests only; no PR code, dependencies or package scripts were executed. Visible CI does not independently establish redaction test coverage.

LangWatch-Review: verdict=findings sha=ec35e9fc3fcec4128870c03379f5ee6a95e93bb4 p0=0 p1=0 p2=1 p3=0

File: packages/redaction/src/secrets.ts [hunk 5]
```diff
@@ -696,9 +708,44 @@ const VALUE_RULES: ValueRule[] = [
...
+      // span, and a slash-containing credential that happens to end in
+      // `sha_…` must keep its protection.
+      const lastSlash = body.lastIndexOf("/");
+      if (lastSlash !== -1) {
+        const tailPrefix = /^([A-Za-z][A-Za-z0-9]{1,11})[_-]/.exec(
+          body.slice(lastSlash + 1),
+        )?.[1];
+        if (
+          tailPrefix &&
+          RECORD_REFERENCE_TAIL_PREFIXES.has(tailPrefix.toLowerCase())
```
[the remark this answers is not in the document]
Comment (plthiyagu) [reply packages/redaction/src/secrets.ts]: Fixed in 78093dbe62. The tail exemption no longer trusts the `so_` prefix: the terminal segment has to match the one shape `deriveStoredObjectId` mints, `^so_0{11}[0-9A-Za-z]{18}$` — a 29-digit base62 KSUID whose fixed zero timestamp (8 zero bytes of the 21-byte payload, so the value is under 2^104 and fits in 18 digits) opens every real id with at least eleven zeros. I confirmed that floor against 20,000 fresh mints through `@langwatch/ksuid` before pinning it; the platform's own `stored-objects.service.unit.test.ts` already asserts `/^so_[0-9A-Za-z]{29}$/`.

`RECORD_ID_PREFIXES` tails keep passing on the prefix alone since this rule never redacts their values anyway; `RECORD_REFERENCE_TAIL_PREFIXES` is gone in favour of an `isRecordReferenceTail` helper that spells out the two cases. Your exact repro (`local-dev-project/so_Ab1Cd2Ef3Gh4Ij5Kl6Mn7Op8Qr9St0`) is the new regression "still redacts it behind a benign path head", verified failing on `ec35e9fc3f`. Package suite 137/137, `tsc` clean.

File: packages/redaction/src/secrets.ts (hunk 4 above)
Comment (plthiyagu) [reply packages/redaction/src/secrets.ts]: Fixed in 78093dbe62, alongside CodeRabbit's sibling finding above: the `so_` tail is now validated as a complete stored-object id (`^so_0{11}[0-9A-Za-z]{18}$`, the shape `deriveStoredObjectId` mints with its fixed zero timestamp) before the reference path is judged, so a terminal prefix is no longer proof. Your counterexample `creds acme_Zx9Qm2Lp7Rt4Vw8s/so_Ke6Ng1Jd5Hf0Cu3Tb9 here` is the `so_` counterpart to the `sha_` test as you suggested ("still redacts a key whose last segment starts with so_ but is no stored-object id"), verified failing on the previous code and passing now; the full 38-char body's judgment stands.

**Review by langwatch-agent — COMMENTED** [review]
No new actionable findings. The previous P2 is addressed: so_ alone no longer exempts the tail, the stored-object ID shape is checked, and both the complete pre-tail body and its segments retain credential-shape checks. Reviewed the new regressions and the ID producer statically. Residual limitation: no external branch code or tests executed; passing visible checks are not independent proof of full redaction-corpus coverage.

LangWatch-Review: verdict=clean sha=78093dbe626dcf0fee08acd16491389752f9993e p0=0 p1=0 p2=0 p3=0
